alf/environments/carla_controller.py

```diff
@@ -0,0 +1,195 @@
+        """
+        self._vehicle = vehicle
+        self._speed_controller = PIDController(s_P, s_I, s_D, step_time)
+        self._direction_controller = PIDController(d_P, d_I, d_D, step_time)
+        self._prev_steering = 0.
+        self._max_speed = max_speed
+        self._max_throttle = max_throttle
+        self._max_steering = max_steering
+        self._max_brake = max_brake
+
+    def action_spec(self):
+        """Get the action spec.
+
+        The action is a 3-D vector of [speed, direction, reverse], where speed is in
+        [-1.0, 1.0] with negative value meaning zero speed and 1.0 corresponding
+        to maximally allowed speed as provided by the ``max_speed`` argument for
+        ``__init__()``, and direction is the relative direction with 0 being
+        front, -0.5 being left and 0.5 being rignt, and reverse is interpreted
+        as a boolean value with values greater than 0.5 corrsponding to True.
+
+        Returns:
+            alf.BoundedTensorSpec
+        """
+        return alf.BoundedTensorSpec([3],
+                                     minimum=[-1., -1., 0.],
+                                     maximum=[1., 1., 1.])
+
+    def action_desc(self):
+        """Get the description about the action.
+
+        Returns:
+            str: the description about the action
+        """
+        return (
+            "3-D vector of [speed, direction, reverse], where speed is in "
+            "[-1.0, 1.0] with negative value meaning zero speed and 1.0 corresponding to "
+            "maximally allowed speed as provided by max_speed argument for __init__(), "
+            "direction is the relative direction with 0 being front, -0.5 "
+            "being left and 0.5 being rignt, and reverse is interpreted as a "
+            "boolean value with values greater than 0.5 corrsponding to True.")
+
+    def act(self, action):
+        """Generate carla.VehicleControl based on ``action``
+
+        Args:
+            action (np.ndarray): 3-D vector representing action
+        Returns:
+            carla.VehicleControl
+        """
+        target_speed, target_direction, reverse = action
+        target_speed = max(0., target_speed) * self._max_speed
+        target_direction = min(1., max(-1., target_direction))
+        target_direction = math.pi * target_direction
+        v = self._vehicle.get_velocity()
+        speed = math.sqrt(v.x * v.x + v.y * v.y + v.z * v.z)
+
+        control = carla.VehicleControl()
+
+        accelaration = self._speed_controller.step(speed, target_speed)
+        if accelaration >= 0.:
+            control.throttle = min(accelaration, self._max_throttle)
+            control.brake = 0.
+        else:
+            control.brake = min(-accelaration, self._max_brake)
+            control.throttle = 0.
+
+        reverse = bool(reverse > 0.5)
+        if reverse:
+            target_direction = -target_direction
```

Comment:
It might be better to add comments in ```action_spec``` and ```action_desc``` to explain that  ```front direction``` corresponding to backwards when reverse is ```True```.